### DISCUSSION TITLE
Decorating in an encapsulated context

### DISCUSSION THREAD

**Discussion by kraftwerk28** [fastify/fastify#4041] [Q&A]
```javascript
import fastify from 'fastify';
import fastifyPlugin from 'fastify-plugin';
import pino from 'pino';

const logger = pino();
const app = fastify({ logger });

const plugin1 = fastifyPlugin(async (app) => {
  app.decorateRequest('first', { getter: () => 'bar' });
});

const plugin2 = fastifyPlugin(async (app) => {
  app.decorateRequest('second', { getter: () => 'baz' });
});

app.register(plugin1);

app.register(async (app) => {
  app.register(plugin2);
  app.post('/', (req, res) => res.send([req.first, req.second])); // gives `[null, "baz"]`
});

app.listen({ port: process.env.PORT });
```

Is this the expected behavior? I noticed that only top-level **request** decorators get vanished; when using `decorate` (and `res.send([app.first, app.second])`) or `decorateReply` (and `res.send([res.first, res.second])`), the properties are preserved from both plugins.

**Comment by jsumners** [comment]
Have you read through https://www.fastify.io/docs/latest/Reference/Encapsulation/?

**Reply by kraftwerk28** [reply]
Yes, is there anything that my snippet is not aware of? I think I faced the bug that's only present in `fastify@4.0.0-rc.1..3`, where `decorateRequest` does not create the prototype chain properly. Any version after these don't have this bug.

**Reply by kraftwerk28** [reply]
Yes, here it is: https://github.com/fastify/fastify/pull/3954, sorry for bothering you.

**Reply by jsumners** [reply]
So your provided code works correctly on the final released versions of Fastify 4?

**Reply by kraftwerk28** [reply]
Yes, it works

**Reply by jsumners** [reply]
👍
